core/callable.cpp

```diff
@@ -0,0 +1,357 @@
+			if (eq_a == eq_b) {
+				return eq_a(custom, p_callable.custom);
+			} else {
+				return false;
+			}
+		} else {
+			return object == p_callable.object && method == p_callable.method;
+		}
+	} else {
+		return false;
+	}
+}
+bool Callable::operator!=(const Callable &p_callable) const {
+	return !(*this == p_callable);
+}
+bool Callable::operator<(const Callable &p_callable) const {
+	bool custom_a = is_custom();
+	bool custom_b = p_callable.is_custom();
+
+	if (custom_a == custom_b) {
+		if (custom_a) {
+			if (custom == p_callable.custom) {
+				return false; //same pointer, dont even compare
+			}
+
+			CallableCustom::CompareLessFunc less_a = custom->get_compare_less_func();
+			CallableCustom::CompareLessFunc less_b = p_callable.custom->get_compare_less_func();
+			if (less_a == less_b) {
+				return less_a(custom, p_callable.custom);
+			} else {
+				return less_a < less_b; //it's something..
+			}
+
+		} else {
+			if (object == p_callable.object) {
+				return method < p_callable.method;
+			} else {
+				return object < p_callable.object;
+			}
+		}
+	} else {
+		return int(custom_a ? 1 : 0) < int(custom_b ? 1 : 0);
+	}
+}
+
+void Callable::operator=(const Callable &p_callable) {
+	if (is_custom()) {
+		if (p_callable.is_custom()) {
+			if (custom == p_callable.custom) {
+				return;
+			}
+		}
+
+		if (custom->ref_count.unref()) {
+			memdelete(custom);
+		}
+	}
+
+	if (p_callable.is_custom()) {
+		method = StringName();
+		if (!p_callable.custom->ref_count.ref()) {
+			object = 0;
+		} else {
+			object = 0;
+			custom = p_callable.custom;
+		}
+	} else {
+		method = p_callable.method;
+		object = p_callable.object;
+	}
+}
+
+Callable::operator String() const {
+
+	if (is_custom()) {
+		return custom->get_as_text();
+	} else {
+		if (is_null()) {
+			return "null::null";
+		}
+
+		Object *base = get_object();
+		if (base) {
+			String class_name = base->get_class();
+			Ref<Script> script = base->get_script();
+			if (script.is_valid() && script->get_path().is_resource_file()) {
+
+				class_name += "(" + script->get_path().get_file() + ")";
+			}
+			return class_name + "::" + String(method);
+		} else {
+			return "null::" + String(method);
+		}
+	}
+}
+
+Callable::Callable(const Object *p_object, const StringName &p_method) {
+	if (p_method == StringName()) {
+		object = 0;
+		ERR_FAIL_MSG("Method argument to Callable constructor must be a non-empty string");
+	}
+	if (p_object == nullptr) {
+		object = 0;
+		ERR_FAIL_MSG("Object argument to Callable constructor must be non-null");
+	}
+
+	object = p_object->get_instance_id();
+	method = p_method;
+}
+
+Callable::Callable(ObjectID p_object, const StringName &p_method) {
+	if (p_method == StringName()) {
+		object = 0;
+		ERR_FAIL_MSG("Method argument to Callable constructor must be a non-empty string");
+	}
+
+	object = p_object;
+	method = p_method;
+}
+Callable::Callable(CallableCustom *p_custom) {
+	if (p_custom->referenced) {
```

Comment:
I don't understand this limitation. Why you can't create two Callable with the same custom? The equality function even try to compare them by pointer.